projects/chaos/src/gpio_fsm.c

```diff
@@ -0,0 +1,101 @@
+#include "gpio_fsm.h"
+
+#include <stdbool.h>
+
+#include "chaos_config.h"
+#include "chaos_events.h"
+#include "event_queue.h"
+#include "fsm.h"
+#include "gpio.h"
+#include "gpio_seq.h"
+#include "misc.h"
+
+#define GPIO_FSM_SLEW_RATE 10
+
+static FSM s_gpio_fsm;
+
+FSM_DECLARE_STATE(idle);
+FSM_DECLARE_STATE(charge);
+FSM_DECLARE_STATE(drive);
+
+FSM_STATE_TRANSITION(idle) {
+  FSM_ADD_TRANSITION(CHAOS_EVENT_GPIO_CHARGE, charge);
+  FSM_ADD_TRANSITION(CHAOS_EVENT_GPIO_DRIVE, drive);
+}
+
+FSM_STATE_TRANSITION(charge) {
+  FSM_ADD_TRANSITION(CHAOS_EVENT_GPIO_IDLE, idle);
+}
+
+FSM_STATE_TRANSITION(drive) {
+  FSM_ADD_TRANSITION(CHAOS_EVENT_GPIO_IDLE, idle);
+}
+
+static void prv_gpio_state_idle(FSM *fsm, const Event *e, void *context) {
+  const ChaosConfig *cfg = context;
+  const GPIOAddress sequence[] = {
+    cfg->motor_interface_power, cfg->rear_camera_power,    cfg->front_lights_power,
+    cfg->rear_lights_power,     cfg->driver_display_power, cfg->array_sense_power,
+    cfg->battery_box_power,
+  };
+
+  gpio_seq_set_state(sequence, SIZEOF_ARRAY(sequence), GPIO_STATE_LOW, GPIO_FSM_SLEW_RATE);
+}
+
+static void prv_gpio_state_charge(FSM *fsm, const Event *e, void *context) {
+  const ChaosConfig *cfg = context;
+  const GPIOAddress sequence[] = {
+    cfg->driver_display_power,
```

Comment:
Hmm, we'll also want to decide on a front array power rail. We should probably make sure to remember that.